pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java

```diff
@@ -379,9 +402,19 @@ public CallableStatement prepareCall(String sql) throws SQLException {
   @Override
   public Map<String, Class<?>> getTypeMap() throws SQLException {
     checkClosed();
+    return new HashMap<String, Class<?>>(getTypeMapNoCopy());
```

Comment:
What's the point of adding a copy here?

Reply:
Per https://docs.oracle.com/javase/8/docs/api/java/sql/Connection.html#getTypeMap--

"You must invoke setTypeMap after making changes to the Map object returned from getTypeMap as a JDBC driver may create an internal copy of the Map object passed to setTypeMap"

This implies that the returned Map must be modifiable.

And the tutorial shows the use-case of it being always modifiable:
https://docs.oracle.com/javase/tutorial/jdbc/basics/sqlcustommapping.html#using_connection_type_map

Reply:
What about using creating a copy on setTypeMap using unmodifiableMap ?
Note it does say "may" create an internal copy, candidly I presume anyone making a change would call setTypeMap so it's sort of pointless, but if we do it on the setter it gets rid of all the internal copying